---
title: "equalizer Vignette"
author: "David Quigley"
date: "May 19, 2016"
output: pdf_document
---


Summary
=========

Affymetrix microarrays are designed using a reference genome from human, mouse, or another organism. Single Nucleotide Polymorphisms (SNPs) in the genomes of individual humans and model organisms that vary from the reference annotations affect the binding of cDNA to microarray probes, affecting the results of experiments measuring gene expression in organisms with distinct genomes. Genetic experiments performed using these arrays are subject to systemic bias unless the effect of these SNPs is accounted for.

The software package equalizer drastically reduces this problem for the commonly used Affymetrix IVT and Gene ST platforms by using whole-genome sequence data to remove probes which overlap SNPs. The customized annotation files generated by equalizer can be used for normalization in oligo or other packages. 

Demonstration of equalizer
=============================

Goal
--------

Run *equalizer* on Affymetrix HG-U133_PLUS_2 chips, creating a file set that corrects for SNPs found in dbSNP build b147.

The dbSNP build 147 VCF file can currently be downloaded at  
*ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b147_GRCh37p13/VCF/common_all_20160408.vcf.gz*. 

Commands to be run from the command line shell are preceeded by **SHELL**.  
Code will be executed from the current working directory.  
Annotation files will be stored in a subdirectory *annotation*.  
Output files will be stored in a subdirectory *output*.  

Step 1. install BedTools 2.17. Get findpython source code.
---------------------------------------------------------------------------

(you can skip this if BedTools is already present and findpython is already installed)

**SHELL**:
```
mkdir output
mkdir annotation

# Obtain and install Bedtools
wget https://bedtools.googlecode.com/files/BEDTools.v2.17.0.tar.gz
tar -xzf BEDTools.v2.17.0.tar.gz
cd bedtools-2.17.0
sudo make 
cd ..

# obtain findpython package for use in R
wget https://cran.r-project.org/src/contrib/findpython_1.0.1.tar.gz
```

Step 2. Obtain dbSNP VCF file and split it into manageable chunks 
-----------------------------------------------------------------------

**SHELL**:
```
wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b147_GRCh37p13/VCF/common_all_20160408.vcf.gz
gunzip common_all_20160408.vcf.gz
mv common_all_20160408.vcf annotation
```

**Setback: The VCF file is very large**

*common_all_20160408.vcf* is so large (35,864,760 elements) that bedtools tends to crash when parsing it. To get around this, one can separate the large single VCF file into one file for each chromosome using a script (*split_vcf_by_chr.py*) that breaks a VCF file up by chromosome. The script is available on my github account, and takes as parameters the input VCF file name and the output file base; for a file base *my_big_file*, files *my_big_file_1.vcf*, *my_big_file_2.vcf*, etc... will be created.

**SHELL**:
```
ROOT_URL=https://raw.githubusercontent.com/DavidQuigley/QuantitativeGenetics
FILE_PATH=/master/equalizer/exec/split_vcf_by_chr.py
wget ${ROOT_URL}${FILE_PATH}
python split_vcf_by_chr.py \
   -f annotation/common_all_20160408.vcf \
   -o annotation/common_all_20160408
```

The result of this is 21 additional files in the annotation directory; we will pass these as VCF files into *equalizer*.

Step 3: Obtain Affymetrix annotation files to alter
--------------------------------------------------------------------------------

**SHELL**:
```
cd annotation
wget http://davidquigley.com/software/equalizer/affymetrix_HG-U133_PLUS_2_annotation_files.tar.gz
gunzip affymetrix_HG-U133_PLUS_2_annotation_files.tar.gz 
tar -xf affymetrix_HG-U133_PLUS_2_annotation_files.tar
AFFY_DIR=http://media.affymetrix.com/analysis/downloads/expression_array_bed_files
wget ${AFFY_DIR}/HG-U133_Plus_2.hg19.bed.gz
gunzip HG-U133_Plus_2.hg19.bed.gz
cd ..
```

Step 4: Synchronize the chromosome annotations
--------------------------------------------------------------------------------

There are two ways to write "chromosome 6" in VCF and BED files: either "chr6" or just "6". Both the probe bed and VCF files need to agree on the format. The Affymetrix probe BED file *HG-U133_Plus_2.hg19.bed* uses the "chr6" format, while the dbSNP VCF file does *not*, so they will disagree. If you run *equalizer* at this point, it will complain that the VCF and BED files have different chromosome formats.

To resolve this legitimate complaint, *HG-U133_Plus_2.hg19.bed* can be re-written in the "no chr" format using a 
simple script *remove_chr_from_vcf.py* that I provide on github:

**SHELL**:
```
ROOT_URL=https://raw.githubusercontent.com/DavidQuigley/QuantitativeGenetics
FILE_PATH=/master/equalizer/exec/remove_chr_from_vcf.py
wget ${ROOT_URL}${FILE_PATH}
python remove_chr_from_vcf.py \
   -f annotation/HG-U133_Plus_2.hg19.bed \
   -o annotation/HG-U133_Plus_2.hg19_nochr.bed
```

This will create a new file, *HG-U133_Plus_2.hg19_nochr.bed*, that should be used in place of the original.

Step 5: Install equalizer and dependencies
--------------------------------------------------------------------------------

(You can skip this if *equalizer* is already installed)

**R**:

```{r eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite()
biocLite("pdInfoBuilder")
install.packages("findpython_1.0.1.tar.gz", repos = NULL, type="source")
install.packages("http://davidquigley.com/software/equalizer/equalizer_0.3.1.tar.gz")
```

Step 6: Run *equalizer* on HG-U133 plus 2.0
--------------------------------------------------------------------------------

The *equalizer* functions *equalize_IVT()* and *equalize_gene()* require a long list of parameters; see the documentation by calling *?equalize_IVT*.

A few notes:  

* *package_name* is the name of the new package you wish to create; since I am amending hgu133plus2 with dbSNPb147, I named the new package hgu133plus2_dbSNPb147.
* I usually supply a new value for *annot_revision* because the annotation I am using is customized.

**R**

```{r eval=FALSE}
library(findpython)
library(equalizer)

fn_python = find_python_cmd()
dir_bedtools_bin = 'bedtools-2.17.0/bin'

package_name="hgu133plus2_dbSNPb147"
# make the list of VCF files, one for each chromosome
chroms = 1:21
fn_vcf_list = paste("annotation/common_all_20160408_",chroms,".vcf", sep="")
affy_bed="annotation/HG-U133_Plus_2.hg19_nochr.bed"
annot_version="na34"
annot_revision="1.1"
genome_version_UCSC="hg19"
genome_version_NCBI="GRCh37"
chip = "HG-U133_Plus_2"
chip_version = "1.1"
species="Homo sapiens"
organism="Human"
author="David Quigley"
email="David.Quigley@ucsf.edu"
fn_CDF='annotation/HG-U133_Plus_2.cdf'
fn_probe_tab = 'annotation/HG-U133_Plus_2.probe_tab'
fn_CEL = 'annotation/example_HG-U133_PLUS_2.CEL'
fn_probeset_csv = 'annotation/HG-U133_Plus_2.na34.annot.csv'
dir_out = "output"


equalize_IVT(
        fn_python, dir_bedtools_bin,
        package_name, fn_vcf_list, affy_bed,
        annot_version, annot_revision,
        genome_version_UCSC, genome_version_NCBI,
        chip, chip_version, species, organism,
        author, email,
        fn_CDF, fn_probe_tab, fn_CEL, fn_probeset_csv,
        dir_out)


pc = read.table('output/probe_count_changes.txt', header=TRUE, stringsAsFactors=FALSE)
length(unique(pc$probeset))
#[1] 54675

length(unique(pc$probeset[pc$probe_removed]))
#[1] 41708
length(unique(pc$probeset[!pc$probe_removed]))

dim(pc)
sum(pc$probe_removed)
#119587
sum(!pc$probe_removed)
#484671
```

This produces the following output, idenfiying SNPs in 41708 probes.

```
*** EQUALIZER version 0.3
*** Copyright 2015 David Quigley
*** Contact David Quigley (dquigley@cc.ucsf.edu, davidquigley.com)
*** Released June 05 2015
MESSAGE: New package will be named: hgu133plus2_dbSNPb147
MESSAGE: New files will be written to: output
MESSAGE: Affymetrix probe sequence file: annotation/HG-U133_Plus_2.hg19_nochr.bed
MESSAGE: User passed the following VCF files to match against probes:
         annotation/common_all_20160408_1.vcf
         annotation/common_all_20160408_2.vcf
         annotation/common_all_20160408_3.vcf
         annotation/common_all_20160408_4.vcf
         annotation/common_all_20160408_5.vcf
         annotation/common_all_20160408_6.vcf
         annotation/common_all_20160408_7.vcf
         annotation/common_all_20160408_8.vcf
         annotation/common_all_20160408_9.vcf
         annotation/common_all_20160408_10.vcf
         annotation/common_all_20160408_11.vcf
         annotation/common_all_20160408_12.vcf
         annotation/common_all_20160408_13.vcf
         annotation/common_all_20160408_14.vcf
         annotation/common_all_20160408_15.vcf
         annotation/common_all_20160408_16.vcf
         annotation/common_all_20160408_17.vcf
         annotation/common_all_20160408_18.vcf
         annotation/common_all_20160408_19.vcf
         annotation/common_all_20160408_20.vcf
         annotation/common_all_20160408_21.vcf
MESSAGE: User indicated chip was HG-U133_Plus_2, version 1.1 annotation version na34
MESSAGE: Does Affymetrix bed file use 'chr' format? False
MESSAGE: Executing BEDtools commands to combine VCF files (combine_vcf.sh)
MESSAGE: Please be patient, as this can take several minutes...

MESSAGE: processing annotation/common_all_20160408_1.vcf
MESSAGE: processing annotation/common_all_20160408_2.vcf
MESSAGE: processing annotation/common_all_20160408_3.vcf
MESSAGE: processing annotation/common_all_20160408_4.vcf
MESSAGE: processing annotation/common_all_20160408_5.vcf
MESSAGE: processing annotation/common_all_20160408_6.vcf
MESSAGE: processing annotation/common_all_20160408_7.vcf
MESSAGE: processing annotation/common_all_20160408_8.vcf
MESSAGE: processing annotation/common_all_20160408_9.vcf
MESSAGE: processing annotation/common_all_20160408_10.vcf
MESSAGE: processing annotation/common_all_20160408_11.vcf
MESSAGE: processing annotation/common_all_20160408_12.vcf
MESSAGE: processing annotation/common_all_20160408_13.vcf
MESSAGE: processing annotation/common_all_20160408_14.vcf
MESSAGE: processing annotation/common_all_20160408_15.vcf
MESSAGE: processing annotation/common_all_20160408_16.vcf
MESSAGE: processing annotation/common_all_20160408_17.vcf
MESSAGE: processing annotation/common_all_20160408_18.vcf
MESSAGE: processing annotation/common_all_20160408_19.vcf
MESSAGE: processing annotation/common_all_20160408_20.vcf
MESSAGE: processing annotation/common_all_20160408_21.vcf
MESSAGE: Reading probesets with SNPs file:
         output/combined.bed
MESSAGE: VCF file analysis identified 41708 probes containing SNPs.
MESSAGE: Reading CDF file:
         annotation/HG-U133_Plus_2.cdf
MESSAGE: Wrote R script to create package: output/create_package.R
MESSAGE: There are several ways to run this script:
         1) Open the file, run R, and paste the contents into R
         2) From the command line, type: 
            R < output/create_package.R --no-save
 
         The script will attempt to install the bioconductor 'pdInfoBuilder' package if 
         it is not already present on your machine. If you do not have the ability to 
         install packages on your machine, contact your local system administrator.
         To install the new package, uncomment and run the last line of the R script.
         This line reads:
         #install.packages('output/pd.hgu133plus2.dbsnpb147', repos = NULL, type='source')
```

